store/sqlstore/sqlx_wrapper.go

```diff
@@ -0,0 +1,342 @@
+	defer cancel()
+
+	var then time.Time
+	if w.trace {
+		then = time.Now()
+	}
+
+	res, err := w.Tx.NamedExecContext(ctx, query, arg)
+
+	if w.trace {
+		printArgs(query, time.Since(then), arg)
+	}
+	return res, err
+}
+
+func (w *sqlxTxWrapper) NamedQuery(query string, arg interface{}) (*sqlx.Rows, error) {
+	if w.Tx.DriverName() == model.DatabaseDriverPostgres {
+		query = namedParamRegex.ReplaceAllStringFunc(query, strings.ToLower)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
+	defer cancel()
+
+	var then time.Time
+	if w.trace {
+		then = time.Now()
+	}
+
+	// There is no tx.NamedQueryContext support in the sqlx API. (https://github.com/jmoiron/sqlx/issues/447)
+	// So we need to implement this ourselves.
+	type result struct {
+		rows *sqlx.Rows
+		err  error
+	}
+
+	// Need to add a buffer of 1 to prevent goroutine leak.
+	resChan := make(chan *result, 1)
+	go func() {
+		rows, err := w.Tx.NamedQuery(query, arg)
+		resChan <- &result{
+			rows: rows,
+			err:  err,
+		}
+	}()
+
+	// staticcheck fails to check that res gets re-assigned later.
+	res := &result{} //nolint:staticcheck
+	select {
+	case res = <-resChan:
+	case <-ctx.Done():
+		res = &result{
+			rows: nil,
+			err:  ctx.Err(),
+		}
+	}
+
+	if w.trace {
+		printArgs(query, time.Since(then), arg)
+	}
+
+	return res.rows, res.err
+}
+
+func (w *sqlxTxWrapper) QueryRowX(query string, args ...interface{}) *sqlx.Row {
+	query = w.Tx.Rebind(query)
+	ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
+	defer cancel()
+
+	var then time.Time
+	if w.trace {
+		then = time.Now()
+	}
+
+	row := w.Tx.QueryRowxContext(ctx, query, args...)
+
+	if w.trace {
+		printArgs(query, time.Since(then), args)
+	}
+
+	return row
+}
+
+func (w *sqlxTxWrapper) QueryX(query string, args ...interface{}) (*sqlx.Rows, error) {
+	query = w.Tx.Rebind(query)
+	ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
+	defer cancel()
+
+	var then time.Time
+	if w.trace {
+		then = time.Now()
+	}
+
+	rows, err := w.Tx.QueryxContext(ctx, query, args)
+
+	if w.trace {
+		printArgs(query, time.Since(then), args)
+	}
+	return rows, err
+}
+
+func (w *sqlxTxWrapper) Select(dest interface{}, query string, args ...interface{}) error {
+	query = w.Tx.Rebind(query)
+	ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
+	defer cancel()
+
+	var then time.Time
+	if w.trace {
+		then = time.Now()
+	}
+
+	err := w.Tx.SelectContext(ctx, dest, query, args...)
+
+	if w.trace {
+		printArgs(query, time.Since(then), args)
+	}
+	return err
+}
+
+func removeSpace(r rune) rune {
+	// Strip everything except ' '
+	// This also strips out more than one space,
+	// but we ignore it for now until someone complains.
+	if unicode.IsSpace(r) && r != ' ' {
+		return -1
+	}
+	return r
+}
+
+func printArgs(query string, dur time.Duration, args ...interface{}) {
+	query = strings.Map(removeSpace, query)
+	switch len(args) {
```

Comment:
Do we really need a `switch` here? Looks like we are only interested in the `len > 0` case. I am thinking that even the duration can be appended to the mlog fields regardless of the case we are in.